map.go

```diff
@@ -1559,10 +1575,17 @@ func (mi *MapIterator) Next(keyOut, valueOut interface{}) bool {
 	if mi.err != nil || mi.done {
 		return false
 	}
+	if mi.drain {
+		return mi.nextDrain(keyOut, valueOut)
+	}
+	return mi.nextIterate(keyOut, valueOut)
+}
 
-	// For array-like maps NextKey returns nil only after maxEntries
-	// iterations.
-	for mi.count <= mi.maxEntries {
+func (mi *MapIterator) nextIterate(keyOut, valueOut interface{}) bool {
+	// For array-like maps NextKey returns nil only after maxEntries iterations.
+	// For maps with keySize equal to 0 (Queue/Stack) we return ErrIterationAborted,
+	// since NextKey returns an error. In this case, Map.Drain should be used instead.
+	for mi.target.keySize != 0 && mi.count <= mi.maxEntries {
```

Comment:
I'm a bit unsure about using keySize like this, it's a very roundabout way of referring to queue / stack. What if there are maps in the future without a key size?

My personal preference would be to drop the special case and make sure that whatever error we get from `Queue.Drain().Next()` due to next key failing is sensible.